### PR TITLE
New Example: Creating reproducible screenshots with a viewer loop

### PR DESCRIPTION
**Pull request by kolibril13**
# Description

This example creates plots in three dimensions from multiple samples.
The special thing about it: 
* The window will open with exact frame width and frame height and the sidepanel will always be closed, that makes these plots reproducible.
* Every plot will have the same angle, same zoom level and same axes and only needs one napari viewer instance. Here is a screen recording of the script:

https://user-images.githubusercontent.com/44469195/149638466-98d57274-d163-4097-bffc-bd643f1e3d8d.mov

This can be e.g. useful when one has dozens of ct scans and wants to visualize them for a quick overview with napari but does not want to load them one by one.

 
## Type of change
- [x] New feature (non-breaking change which adds functionality)

# References

Thanks to @haesleinhuepf @tlambert03  in the image.sc forum who helped me today to realize this idea!
https://forum.image.sc/t/hide-layer-controls-panel-from-python/61975
https://forum.image.sc/t/update-refresh-a-layer-in-napari-julia-napari-jl/59728/9
# How has this been tested?
Runs properly with macOS, python3.8, napari 0.4.12

## Final checklist:
- [x] My PR is the minimum possible work for the desired functionality
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

File: examples/adjustwindow_updateviewer_saveimage.py
Comment (brisvag) [inline examples/adjustwindow_updateviewer_saveimage.py]: 0.4.13 is about to come out (we're currently testing release candidates), so we should definitely avoid suggesting the use of a private attribute in an example! Why is this private call necessary?

Comment (brisvag) [reply examples/adjustwindow_updateviewer_saveimage.py]: Sorry if this came out a bit dry, I meant it as part as a wider review and it fired off as a single comment :) I'll add more when I understood the whole thing!

Comment (haesleinhuepf) [reply examples/adjustwindow_updateviewer_saveimage.py]: > 0.4.13 is about to come out (we're currently testing release candidates), so we should definitely avoid suggesting the use of a private attribute in an example! Why is this private call necessary?

See related discussion: https://forum.image.sc/t/hide-layer-controls-panel-from-python/61975/8?u=haesleinhuepf

Comment (brisvag) [reply examples/adjustwindow_updateviewer_saveimage.py]: Hah, it's been discussed already! Thanks for the link @haesleinhuepf. Ok then! :+1:

Comment (tlambert03) [reply examples/adjustwindow_updateviewer_saveimage.py]: > Why is this private call necessary?

Essentially, we're trying to make *all* of the actual Qt API a napari application implementation detail, and not part of our public API.  Of course, there's a ton of things an end-user might want to do _with_ Qt (and this example is using a Qt API directly).

It's a difficult thing to decide and will definitely require some discussion in the future.  If we just say "go ahead, use any of our Qt objects, here's the Qt API", then people can do anything, and that's great, but we can't evolve at all without breaking a ton of stuff.

If we make it private, people at least know that we're not guaranteeing we won't change it (and that plugins definitely shouldn't use it).  But end-users can still access "riskily" :)

I don't have a super strong feeling for a "rule" on how we should handle this, so, for now, I tend to just say "here's how ya do it, but, just a heads up here...".  And if we get a couple requests to make a public API for something, we'll know it's important.  (on the other hand, we don't want to just forward the entire Qt API, so some decision making is necessary)

Comment (kolibril13) [reply examples/adjustwindow_updateviewer_saveimage.py]: From my user perspective:
> but we can't evolve at all without breaking a ton of stuff.

I think breaking stuff to evolve is a good strategy, adjusting my user scripts to new version is not a big deal.

Regarding this special case of showing and hiding the side panel from python side:
I am mostly using napari on my laptop with split-screen: left VSCode, right napari.
As there is only limited space, I want to see more of my data and switch off the side panel manually.
Next,  I change some settings from python, e.g.
```py
viewer.camera.zoom = 1.3
viewer.axes.visible = True
viewer.camera.angles = (90, 27, 145)
```
But these parameters are optimized for the case that the side panel is off.
As soon as I close my program and start it again a few days later, napari will start with the side panel on, hence the viewer has a different aspect ratio in the canvas and therefore my settings that I've set last time won't work anymore.
Of course, I can simply remember that I made these settings with "side panel is off", but then I would have to remember for all my scripts, if I optimized the viewer parameter with  "side panel is off" or  "side panel on".
That is why I think it would be nice if there was an option for users to change this setting by the API.

Comment (tlambert03) [reply examples/adjustwindow_updateviewer_saveimage.py]: > I think breaking stuff to evolve is a good strategy,

I am also someone who doesn't mind updating my scripts when a dependency makes breaking changes, but I think we might be in the minority there.  Or, at the very least, it's not considered a good practice for a package to make unannounced breaking changes.  With public attributes and methods, we pretty much _must_ go through a deprecation process: announcing the change and then later making it.  It makes for a _slower_ change process, and a messier code base.  Private methods announce: "here is something that we reserve the right to change without notice".  If you are the kind of person who doesn't mind updating stuff when it changes and are happy for your libs to make breaking changes, then private methods will work nicely for you :)

Comment (tlambert03) [reply examples/adjustwindow_updateviewer_saveimage.py]: one more thought: this particular case brings up what might be a good rule of thumb.  Anything that has such an easy thing to do via the menus (here: toggling an item in the `Window` menu) should probably have a corresponding public method

Comment (kolibril13) [reply examples/adjustwindow_updateviewer_saveimage.py]: One of the first things in the GitHub README is:
> We're developing napari in the open! But the project is in an **alpha** stage, and there will still likely be **breaking changes** with each release. 

So even when there is no announcement that this method will become private, this breaking change is in my opinion completely fine. 
I like your philosophy of private methods, that is a nice middle course: Giving the developers more freedom for faster development while still preserving the freedom of users to tweak also internal properties.

Regarding your second idea: public methods for Window property would be amazing!
This line was also a bit hacky:
```py
from napari.settings import get_settings
get_settings().application.window_size = (500, 500)
```
Maybe this can also become a public method?

Additional thought: Changing `window_size` has to happen *before* the viewer is initialized, while hiding the side panel has to happen *after* the viewer is initialized. Maybe this can be more coherent, maybe everything can happen after the viewer is initialized?

Comment (tlambert03) [reply examples/adjustwindow_updateviewer_saveimage.py]: > from napari.settings import get_settings; get_settings().application.window_size = (500, 500)
> Maybe this can also become a public method?

that _is_ a public method :)  anything without a leading underscore is public.

Comment (tlambert03) [reply examples/adjustwindow_updateviewer_saveimage.py]: >  Changing window_size has to happen before the viewer is initialized, while hiding the side panel has to happen after the viewer is initialized.

only if you change it in the settings, afterwards, you'd need to use `viewer.window.resize(width, height)`

*edit*: think of `get_settings()` as the _persistent_ settings... not the way you're supposed to change existing objects.  Almost everything the settings affects has a direct way to do it after instantiation.

Comment (kolibril13) [reply examples/adjustwindow_updateviewer_saveimage.py]: > `viewer.window.resize(width, height)`

This is exactly what I was looking for 😍
I'll change this pr accordingly in a few minutes.
I just realized that there is a minimum size that is required.
Maybe one can send the user a warning when the user wants to set one value below a certain threshold here?
https://github.com/napari/napari/blob/1cc0592a02bd7b06c3b01c16128466aea9868494/napari/_qt/qt_main_window.py#L994  I've tried a few values and the minimum value seems to be somewhere at viewer.window.resize(600, 550)

File: examples/adjustwindow_updateviewer_saveimage.py
Comment (kolibril13) [inline examples/adjustwindow_updateviewer_saveimage.py]: @tlambert03 :
I've now replaced `get_settings().application.window_size` with `viewer.window.resize(600, 600)`.
Is there something similar for `get_settings().application.window_position = (900, 300)` ?

Comment (kolibril13) [reply examples/adjustwindow_updateviewer_saveimage.py]: And while I'm on it: Is there also an option in `application.window_position` to define the screen on which napari will open? That would be incredibly useful for a two-screen setup: One screen with the editor, and the viewer would show on the second screen.

Comment (tlambert03) [reply examples/adjustwindow_updateviewer_saveimage.py]: we don't have a public API for that...  you'd have to use the Qt API directly.  I don't have a second monitor to test it out

Comment (sofroniewn) [reply examples/adjustwindow_updateviewer_saveimage.py]: Just catching up here, I'd be open to expanding the API to include a call like `viewer.window.position(600, 600)`. What does @tlambert03 @jni think? It needn't hold up this PR, but I just wanted to add that

File: examples/adjustwindow_updateviewer_saveimage.py
Comment (sofroniewn) [inline examples/adjustwindow_updateviewer_saveimage.py]: hmm we don't have anything else right now that will be set to be a public method like this. I wonder why this is needed though if you have `canvas_only=True` in the screenshot? I'd prefer if we could drop it, but curious what others think

File: examples/adjustwindow_updateviewer_saveimage.py
Comment (sofroniewn) [inline examples/adjustwindow_updateviewer_saveimage.py]: I think when you set the data below you trigger a refresh which might be resetting the camera zoom? I'm not sure, I havn't looked into to it, but we can investigate

**Review by sofroniewn — COMMENTED** [review]
Hi @kolibril13 I just tried to update from main so tests will pass now.

I've left a few comments on your questions above, hopefully we can get them resolved and then get this merged. I wonder if we can maybe shorten/ simplify the name. I wonder also how this example that opens up a new window will interact with our new gallery cc @jni @melissawm? We might have to exclude it for now

I also was actually unable to run this locally now - i got a vispy error

```
RuntimeError: Cannot SIZE object 39 because it does not exist
```

and my screen going nuts with a weird loop.

Does anyone else see this? Maybe @brisvag can investigate this a little?

**Comment by kolibril13** [comment]
Thanks for your review @sofroniewn !

> I wonder if we can maybe shorten/ simplify the name

What about `adjustwindow_updateviewer_saveimage.py` -> `update_viewer_loop.py` ?

> I also was actually unable to run this locally now - i got a vispy error

I am also not able anymore to reproduce this script, a month ago I've also posted this issue here https://forum.image.sc/t/update-refresh-a-layer-in-napari-julia-napari-jl/59728/11

Something must have changed from napari  0.4.12 to napari 0.4.15 that affects the `loop_run logic`, but I think I won't be able to figure that out, maybe @haesleinhuepf has some ideas there.

Two points from my side:
* I just thought it's a fun script to have for reproducible batch processing, but please don't waste too much of your time into it to make it work again, I think it's also totally fine to close this pr, and the discussed points about LayerControls and positioning could be converted into issues. 
* For anyone who wants to go down the rabbit hole nevertheless:  Please feel free to directly commit to my branch.

**Comment by brisvag** [comment]
> ```
> RuntimeError: Cannot SIZE object 39 because it does not exist
> ```

I've seen this before, and it's one of the least useful errors out there xD Could mean anything, as far as I know :P Often it comes with more errors much higher in the traceback that are more informative. Will take a look!

**Comment by brisvag** [comment]
I just tried locally, and everything is fine; are you sure you pulled main, and updated vispy? It might be an issue related with #4401. Try a new env!

I love this example! Shouyld we make it into a jupyter notebook as well? Cause as it is now, I only see one of the figures pop up (not super familiar with how showing figures behaves in ipython, so I might be doing something wrong...)

**Comment by kolibril13** [comment]
> I just tried locally, and everything is fine; are you sure you pulled main, and updated vispy? It might be an issue related with https://github.com/napari/napari/pull/4401. Try a new env!

That's magic, just ran the example again, and it works now properly:
image


> I love this example! Shouyld we make it into a jupyter notebook as well? Cause as it is now, I only see one of the figures pop up (not super familiar with how showing figures behaves in ipython, so I might be doing something wrong...)

That is a good idea, should I change this file to a notebook?
When running this as a python script, I also get only the figures pop up.

File: examples/adjustwindow_updateviewer_saveimage.py
Comment (kolibril13) [reply examples/adjustwindow_updateviewer_saveimage.py]: My three cents what would be lovely to have in the user API:
* An option to resize the window: `viewer.window.resize(600, 600) `  (already there 🎉)
* An option to position the window, e.g. `viewer.window.position(600, 600)` (currently possible with `get_settings().application.window_position = (900, 300)`)
* An option to select the screen where napari will pop up, this would be amazing to have in a multiscreen setup. (I've made an issue about that idea here: https://github.com/napari/napari/issues/3987)

File: examples/adjustwindow_updateviewer_saveimage.py
Comment (brisvag) [reply examples/adjustwindow_updateviewer_saveimage.py]: @kolibril13 and @sofroniewn. note that with #4201 this shouls all be easier!

**Comment by brisvag** [comment]
> What about `adjustwindow_updateviewer_saveimage.py` -> `update_viewer_loop.py` ?

I would suggest something like `reproducible_screenshots` to better reflect the intent/use case. What do you think?

File: examples/adjustwindow_updateviewer_saveimage.py
Comment (kolibril13) [reply examples/adjustwindow_updateviewer_saveimage.py]: I've just tested this code again after three months, and the zoom level does actually not reset anymore.
So I can use this line here :)

File: examples/adjustwindow_updateviewer_saveimage.py
Comment (kolibril13) [reply examples/adjustwindow_updateviewer_saveimage.py]: I changed now the syntax to 
```py
viewer.window._qt_viewer.dockLayerControls.toggleViewAction().trigger()
viewer.window._qt_viewer.dockLayerList.toggleViewAction().trigger()
```
It is not pretty, but it does what it should do.
I don't want to use `canvas_only=True` in this case, because I want to keep the aspect ratio that I've defined in `viewer.window.resize(600, 600)`.

**Comment by kolibril13** [comment]
In the next commit, I'll change it to `viewer_loop_reproducible_screenshots.ipynb` now clean up the code.

File: examples/adjustwindow_updateviewer_saveimage.py
Comment (brisvag) [reply examples/adjustwindow_updateviewer_saveimage.py]: You can sidestepp all of this (toggling controls + setting window size) by using the new `size` argument from #4201!

**Comment by kolibril13** [comment]
You can find the rendered notebook on my branch: https://github.com/kolibril13/napari/blob/update_window_example/examples/viewer_loop_reproducible_screenshots.ipynb
At this place it's better formatted than in the code diff.

File: examples/adjustwindow_updateviewer_saveimage.py
Comment (kolibril13) [reply examples/adjustwindow_updateviewer_saveimage.py]: The `size` feature in `screenshot` is a really cool feature!
I currently installed  `napari 0.4.15` via conda, so I can not use this feature yet, and I don't have the time right now to set up the latest napari from GitHub.
Also, I think in this case it is better to disable the dockLayer manually, because then I have already a preview of how my screenshot will look like and there wont be any unpredictable surprises. 
I prefer the "What you see is what you get" approach here.

File: examples/adjustwindow_updateviewer_saveimage.py
Comment (jni) [reply examples/adjustwindow_updateviewer_saveimage.py]: I like all three! Monitor selection is tricky because I think that is very platform-specific. For example, in Linux using i3, everything appears as one giant, irregular canvas. So I think that will take a while to get right...

**Comment by jni** [comment]
I would like to avoid committing notebooks directly, as they render poorly in diffs and are difficult to edit directly. @kolibril13 would you mind making this a MyST-nb notebook? You need to install jupytext and then follow the instructions on our [contributing documentation guide](https://napari.org/developers/documentation/index.html).

Other than this though, I agree with the general sentiments that this is a great thing to add to our docs. My personal preference would be to use the `size=` preference — though I agree that wysiwyg is valuable, I don't think I want to promote private APIs in our docs. We could instead use `size=` for now, then make issues to create all the public APIs we need for wysiwyg, including:

- viewer.window.position
- `viewer.controls.visible = False` (for example)
- `viewer.layers.visible = False` (for example)

But, at least the private APIs are *clearly* private, so I'm not *too* fussed.

**Comment by kolibril13** [comment]
@jni: Thanks for your suggestions!
I agree, the notebook diff looks poor, and I'll convert it to MYST-nb.
Note that there is already another notebook in the napari example folder:
https://github.com/napari/napari/blob/main/examples/notebook.ipynb , I can convert this as well and add it to this pull request.
I really like wysiwyg, therefore I will change the parameter to `canvas_only=True` and then remove the calls from the private API that hides the sidepannel, the `size=` parameter is not required then. As soon as there will probably be public APIs (I like the syntax that you suggest!)  this example can be updated.
I think I'll find some time to incorporate these changes in a few hours.

**Comment by jni** [comment]
> I can convert this as well and add it to this pull request.

No, it's best to keep pull requests focused on one thing. Adding extra things can trigger new discussion and block unrelated things from getting merged. Please submit a different PR for that fix, if you like. :blush:

**Comment by kolibril13** [comment]
@jni Side Note: One can have a nice diff of Jupyter notebooks in the .ipynb format in the github VSCodeview view, see
https://github.dev/napari/napari/pull/3947
image
Only drawback: It's not possible to add line comments there.
One issue that I can see with MyST-nb: Everyone who wants to use this notebook in Jupyter has to install jupytext as well.
Therefore: Should I still convert it to MyST-nb?

**Comment by jni** [comment]
Yes, I'm happy to suggest to folks that they should install jupytext, it's an awesome library! 😃 And there are configuration options that allow people to download the ipynb from the website.

**Comment by kolibril13** [comment]
> Yes, I'm happy to suggest to folks that they should install jupytext, it's an awesome library! 😃 And there are configuration options that allow people to download the ipynb from the website.

Nice, then I'll convert it to MyST-nb in a few hours! :)

**Review by sofroniewn — COMMENTED** [review]
Thanks for making those changes @kolibril13.

I wonder how you feel about turning this into more of a how-to-guide instead of just an example? You would move this file so it lives in our how-to-guides https://github.com/napari/napari/tree/main/docs/howtos and add it to our index https://github.com/napari/napari/blob/main/docs/howtos/index.md it will then render on our website here https://napari.org/howtos/index.html.

The how-to-guides also allow you to write a little bit more content about why you are doing something and provide more explanation than is present in just an example. It could maker it easier for new users to understand what you are achieving here.

What do @jni and @brisvag think? I hate to make you reformat this again and again, so maybe let's hear from someone else before making more changes

If you've not already done so would you also mind creating a new issue with the requested public API calls and tag this PR in it.

Thanks again for the work here and your willingness to make changes!!

**Review by jni — APPROVED** [review]

**Comment by jni** [comment]
@sofroniewn

> I wonder how you feel about turning this into more of a how-to-guide instead of just an example? 

That's a good plan but one that could easily come after merging this, since we've already spent a bunch of time iterating on this.

Having said this I also agree that, given that this is a notebook anyway, we could/should have some cells explaining what's going on. What do you think @kolibril13?

**Comment by kolibril13** [comment]
> Having said this I also agree that, given that this is a notebook anyway, we could/should have some cells explaining what's going on. What do you think @kolibril13?

Good idea, I've just added some explaining lines to each cell.
@jni : In case that you find typos, or want to rephrase certain lines, feel free to directly commit your changes to this branch.

> I wonder how you feel about turning this into more of a how-to-guide instead of just an example? 

I think right now I'd like to just have it in the example section, just as @jni suggested.
For making it a How-To-Guide, I don't have the capacity right now, but maybe some time in future I'll come back to this.

> If you've not already done so would you also mind creating a new issue with the requested public API calls and tag this PR in it.

I can take care of this!

**Review by sofroniewn — APPROVED** [review]
Thanks for adding comments. Test fails are unrelated to this PR. I will merge now - we can always convert to a how-to-guide later.

**Pull request merged**
